### PR TITLE
Add Third-Party Ruy Submodule

### PR DESCRIPTION
**Pull request by salilsdesai**
Ruy will be used for quantized matmul support

Commit created with
` git submodule add https://github.com/google/ruy.git third_party/ruy`

### CODE REVIEW

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Do you plan to add build support for this, via cmakes, in a later commit?

**Comment by salilsdesai** [comment]
> Do you plan to add build support for this, via cmakes, in a later commit?

Yes, the instructions on how to add a new third party submodule that I found on a wiki page said to make a pull request like this without any changes to cmake. So I will add the cmake changes in a later PR

**Comment by suo** [comment]
Hi @salilsdesai, as a library we generally like to take a close look at new proposed third-party dependencies, as they can impact users and incur ongoing maintenance costs. I'd like to open a discussion on this dependency; as is I don't think we have enough context on it to determine whether it's merge-able. Some seeding questions for this discussion:
- Why are we adding this dependency?
- Are there any alternatives to adding a third-party dependency to PyTorch core? Can this be done in an out-of-core fashion?
- What are the proposed build options we are adding?
- Will these build options be turned on by default? Do we intend to ship this dependency as part of the default torch binary?
- What is the maintenance status of this library? Do we know it will continue to be maintained in the future? Are we in contact with the maintainers?

…and so on.

**Comment by kimishpatel** [comment]
Hi Michael,
THanks for bringing this up. Please see below.

> * Why are we adding this dependency?

So ruy is library for doing quantized matrix multiplications. Unliked qnnpack/xnnpack which do only stateful ops ruy is stateless. Issue with using qnnpack is that it would incur cost of weight packing if we used it for quantized matmul/bmm. Now once we have quantized matmul/bmm it enables us to have quantized attention module and thus quantized transformers.

> * Are there any alternatives to adding a third-party dependency to PyTorch core? Can this be done in an out-of-core fashion?

One alternative to this is doing our own quantized matmul implementation. We have thought about it and it is possible to do that but with quite a bit of more effort.

With respect to out-of-core question: Can you elaborate on what you mean?

> * What are the proposed build options we are adding?

This is something @salilsdesai is still working on. I would like to know what would be the concerns here so that it is accounted for.

> * Will these build options be turned on by default? Do we intend to ship this dependency as part of the default torch binary?

So the goal is that once we have quantized matmul, quantization workflow will enable us to generate a quantized attention module. (Subsequently we will also have quantized softmax to have entire attention module quantized)

> * What is the maintenance status of this library? Do we know it will continue to be maintained in the future? Are we in contact with the maintainers?
> 

Good question. We are not in contact with the maintainers. We probably can look into this. While this repo is not very active it is still maintained. But if there are concerns about a dependency that is not maintained anymore I would like to understand the concerns. 

> …and so on.

Some questions for you.

What are the things we consider when adding a third party dep?
If we do this for performance reasons, do we have to show proof?
What are generally the reasons we dont want to add third party deps? Build costs, build failures and subsequent maintenance etc?

**Comment by suo** [comment]
The general reason we want to be careful adding third-party dependencies to the core library is that they incur a long-term maintenance cost. Someone needs to make sure it stays updated, any bugs root caused to it find appropriate resolution, that the build continues working, and so on. Cost in $$ terms are not as important as the ongoing human cost of maintaining the dependency.

In the past we have had…mixed results with regard to long-term maintenance. Some dependencies (especially the ones that come with a motivated vendor supporting them) are fine, others are a bit of a headache and the original people who signed up for maintenance have long since disappeared.

I trust the team's judgment on the perf benefits—my major question is whether we can achieve those benefits without adding a new dependency to the core library. For example, can we use our custom operator API to provide this library as a pluggable extension rather than adding it to the core?

Also, is this mostly for internal use cases? Do we expect external users to make use of this submodule? What is the API that they will use for it? These are some other questions worth discussing.

**Comment by kimishpatel** [comment]
@suo some thoughts:

> my major question is whether we can achieve those benefits without adding a new dependency to the core library. For example, can we use our custom operator API to provide this library as a pluggable extension rather than adding it to the core?

> Also, is this mostly for internal use cases? Do we expect external users to make use of this submodule? What is the API that they will use for it? 

I think these two are related so I will try to respond together.

In order to use custom op (or some out of core approach), we will have to introduce out-of-tree module for quantized matmul which depends on pytorch as well as this lib. This way we can register custom op. But the challenge then is integrating it with quantization workflow. Quantization workflow can also possibly be part of the same out-of-tree module. This will probably be cumbersome but doable. However, I think it defeats the purpose. I believe we do want to enable quantized attention module in our quantization workflow that is part of the core quantization workflow rather than custom pluggable extension. Maybe @vkuzo can speak more to it.

But at the moment this is mostly for internal use case, however I want to separate the concern raised here. We can address your in this way:
1. Introduce quantized matmul/bmm that is natively implementd: Dequantize-matmul-quantize.
2. Introduce quantized softmax (to come later) in similar fashion.

Then for internal use case we can do one the two:
1. Introduce JIT pass to rewrite quantized matmuls with ruy based quantized matmul (a custom internal op) or
2. Instead of building OSS version of quantized matmul, build internal optimized one.

For now I am ok with taking the approach where we dont have to introduce ruy to OSS. As we evaluate performance and figure out longer term plan for quantized attention modules (e.g. our native implementation optimized fused matmul+softmax), we can revisit this.

Let me know what you think.

**Comment by vkuzo** [comment]
@kimishpatel , the quantization workflow does not have to be aware of this op if this has to be out of tree.  It could work like this:
1. quantization workflow provides a reference model with bmm_int8 implemented as dequant -> bmm_fp32 -> quant
2. a lowering step (this can be out of tree) lowers the dequant -> bmm_fp32 -> quant pattern to a real kernel

This is the approach we are taking with other backends like TensorRT, it is an option here as well.

**Comment by kimishpatel** [comment]
> @kimishpatel , the quantization workflow does not have to be aware of this op if this has to be out of tree. It could work like this:
> 
> 1. quantization workflow provides a reference model with bmm_int8 implemented as dequant -> bmm_fp32 -> quant
> 2. a lowering step (this can be out of tree) lowers the dequant -> bmm_fp32 -> quant pattern to a real kernel
> 
> This is the approach we are taking with other backends like TensorRT, it is an option here as well.

Yes. I mentioned kind of same approach as one option but this fits more inline with the general theme quantization team is working with. Only thing is that the out-of-box perf will suck and it can discourage users from quantizing attention modules. I dont know if this is something you would want to consider, even if not addressing but at least bringing user awareness to it.

**Comment by suo** [comment]
Yeah, I am totally sympathetic to the idea that we want to provide good "out-of-the-box" perf to users. We definitely don't want to be stuck in a long-term situation where Meta workloads had a lot of nice optimization done but the library that we ship to open source users is super slow.

I think the the overall plan you propose makes sense to me: we do out of core for now, build some empirical experience with ruy and the perf/figure out the long term plan for quantized attention, and we set some time where we are committed to re-evaluating things.

When do you think is a good time? To me, it could either be a fixed period (6mo from now?) or demand driven (for example, if we want to make a bigger push on OSS for this stuff). What do you think?

**Comment by kimishpatel** [comment]
In terms of ETA let's revisit this at the end of Q1 (end of March).

**Comment by salilsdesai** [comment]
Ok sounds good. I'm going to close this pull request for now.
